firehose-writer/src/main/java/com/expedia/www/haystack/pipes/firehoseWriter/FirehoseProcessor.java

```diff
@@ -16,186 +16,84 @@
-        /* do nothing */
+        this.s3Sender = s3Sender;
+        firehoseProcessorLogger.info(String.format(STARTUP_MESSAGE, firehoseConfigurationProvider.streamname()));
+        this.parallelismSemaphore = factory.createSemaphore(firehoseConfigurationProvider.maxparallelismpershard());
     }
 
     @Override
     public void process(String key, Span span) {
         firehoseCountersAndTimer.incrementRequestCounter();
         firehoseCountersAndTimer.recordSpanArrivalDelta(span);
         final List<Record> records = batch.getRecordList(span);
-        sendRecordsToS3(records);
-    }
-
-    @Override
-    public void close() {
-        final List<Record> records = batch.getRecordListForShutdown();
-        sendRecordsToS3(records);
-    }
-
-    private class RetryCalculator {
-        final int initialRetrySleep;
-        final int maxRetrySleep;
-        int boundedTryCount; // never increments more than one step past the value that causes the exponential backoff
-                             // time calculation to exceed maxRetrySleep; this avoids problems for > 31 tries.
-        private RetryCalculator(int initialRetrySleep, int maxRetrySleep) {
-            this.initialRetrySleep = initialRetrySleep;
-            this.maxRetrySleep = maxRetrySleep;
-        }
-
-        /**
-         * Calculates the number of milliseconds to sleep. The first time this method is called, it returns 0.
-         * The second time it returns 1 * initialRetrySleep, the third time it returns 2 * initialRetrySleep,
-         * the fourth time it returns 4 * initialRetrySleep, the fifth time it returns 8 * initialRetrySleep, etc.
-         * But the returned value is bounded by maxRetrySleep; it will never return a number larger than maxRetrySleep.
-         * @return milliseconds to sleep
-         */
-        private int calculateSleepMillis() {
-            final int sleepMillisPerTryCount = (1 << (boundedTryCount - 1)) * initialRetrySleep;
-            final int sleepMillis;
-            if(sleepMillisPerTryCount > maxRetrySleep) {
-                sleepMillis = maxRetrySleep;
-            } else {
-                sleepMillis = sleepMillisPerTryCount;
-                ++boundedTryCount;
-            }
-            return Math.min(sleepMillis, maxRetrySleep);
-        }
+        processRecordsReinterruptingIfInterrupted(records);
     }
 
-    private void sendRecordsToS3(List<Record> records) {
-        int retryCount = 0;
-        int failureCount;
+    private void processRecords(List<Record> records) throws InterruptedException {
         if (!records.isEmpty()) {
-            final String streamName = firehoseConfigurationProvider.streamname();
-            final int maxRetrySleep = firehoseConfigurationProvider.maxretrysleep();
-            final RetryCalculator retryCalculator = new RetryCalculator(
-                    firehoseConfigurationProvider.initialretrysleep(), maxRetrySleep);
+            parallelismSemaphore.acquire();
```

Comment:
Would it be better to acquire and release semphores in block? I wonder whether passing around these semaphores would result in us not closing the semaphores(because of specific path not being handled) there by starving all the requests.